issue 3 solution

Ahmedborwin opened the pull request:
## Description

Reading CheckedInEvent History to create a list of checked in builders + their ens name

## Additional Information

- [ X] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [ X] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

## Related Issues

_Closes #3_


Your ENS/address: 0x5f2AF68dF96F3e58e1a243F4f83aD4f5D0Ca6029

Ahmedborwin commented:
useScaffoldEventHistory to read CheckedIn event returns error "Contract Not Found" first time around, but then is able to find the contract. Functionality is fine, just an error on the console. Unsure how to deal with it. Any suggestions?

technophile-04 commented:
Hey @Ahmedborwin thanks for taking up the issue 🙌 ! 

I see that we are adding `isConnectedOptimism` ideally we shouldn't do that because we want to show the list of checked-in to all the users(even to those who don't have web3 wallet) so whenever anyone lands on home page of https://batch3.buidlguidl.com/ irrespective of whether the user is connected to wallet / on a wrong chain we want to show it.  This applies in general also people any `read` functions/events should be displayed irrespective of user connecting the wallet

--- 

> read CheckedIn event returns error "Contract Not Found" first time around, but then is able to find the contract. Functionality is fine, just an error on the console.

Yeah, you could just ignore it since it's just logging the error in the console. This happens because on very first render none of the states are initialized and in `useDeployedContractInfo` (internally used by `useSCEventHistory`) hook we return undefined at that stage. 

This is something that we can handle better on SE-2 main repo side, but for this PR nothing to worry about lest leave it as it 🙌

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -1,16 +1,36 @@
 
 const Home: NextPage = () => {
+  const [checkedInBuilders, setCheckedInBuilders] = useState<string[]>([]);
+  const [isConnectedOptimism, setIsConnectedOptimism] = useState<boolean>(false);
+
+  //HOOKS
+
+  const { chain } = usePublicClient();
```
Lets remove this checkout first section of https://github.com/BuidlGuidl/batch3.buidlguidl.com/pull/14#issuecomment-1988806853

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -1,16 +1,36 @@
+    contractName: "BatchRegistry",
+    eventName: "CheckedIn",
+    fromBlock: 116978463n,
+    watch: true,
+    enabled: isConnectedOptimism,
```
We could remove both this, we don't need to `watch` since `CheckedIn` event will change occasionally, adding `watch` will fetch every 30 sec checkout [this](https://docs.scaffoldeth.io/hooks/useScaffoldEventHistory#configuration)  and puts loads on RPC 😅

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -19,6 +39,39 @@ const Home: NextPage = () => {
+    return `${address.substring(0, 5)}...${address.substring(address.length - 4)}`;
+  }
+
+  //creates a list of checked in builders addresses from the event history of the contract
+  //uses Set to make sure there are no duplicated on re-render
+  useEffect(() => {
+    if (!isReadingEventLoading && eventHistory && isConnectedOptimism) {
+      const currentBuildersSet = new Set(checkedInBuilders);
+      const builders = eventHistory
+        .map(e => e.args.builder)
+        .filter(builder => builder !== undefined && !currentBuildersSet.has(builder)) as string[];
+      if (builders.length > 0) {
+        setCheckedInBuilders(currentBuilders => [...currentBuilders, ...builders]);
+      }
+    }
+  }, [isReadingEventLoading, eventHistory, checkedInBuilders, isConnectedOptimism]);
```
We actually don't need separate `checkedInBuilders` state and also we could remove this whole `useEffect` and do something like this : 

```ts
const buildersSet = new Set(eventHistory?.map(e => e.args.builder).filter(Boolean));
const checkedInBuilders = Array.from(buildersSet);
```
`.filter(Boolean)` will remove undefined and null values in above code 

Since `eventHistory` is internally stored using `useState` in `useScaffoldEventHistory` the `checkedInBuilders` will be updated as soon as it changes.

Ahmedborwin replied:
Trying to get my head around this - the code is much cleaner that's for sure!

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -32,6 +85,44 @@ const Home: NextPage = () => {
           </p>
+          <div className="flex flex-col text-center text-md">
+            <div>
+              {!isConnectedOptimism ? (
+                <div className="mt-4 text-xl font-bold"> Please Connect to Optimism To Read List of Builders </div>
```
No more need this check check https://github.com/BuidlGuidl/batch3.buidlguidl.com/pull/14#issuecomment-1988806853

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -32,6 +85,44 @@ const Home: NextPage = () => {
+              ) : (
+                <>
+                  <span className="font-extrabold text-lg">List of Builers</span>
+                  {isReadingEventLoading ? (
```
We could add this check to remove flickering of UI : 

`{isReadingEventLoading || !eventHistory?.length ? (`

Ahmedborwin replied:
Haha it shows up underneath the text for me.. are you viewing from a smaller screen? I will have a look anything and see if the issue is responsivness

Ahmedborwin replied:
Hey @technophile-04  - i get what you mean. I am trying to fix this but I am unable to commit my code. I get the following error:
hooks/scaffold-eth/useScaffoldEventSubscriber.ts(33,5): error TS2322: Type 'Abi' is not assignable to type 'InferContractAbi<Contract<TContractName>>'.
I don't think I touched that hook so not sure what the issue is. Also, I have just committed not 10/15 mins ago without issue (although I had to pull before I could commit) . Any ideas?

technophile-04 commented on packages/nextjs/app/page.tsx:
```diff
@@ -32,6 +85,44 @@ const Home: NextPage = () => {
+                          <tr key={index}>
+                            <td className="border p-2">
+                              <span className="hidden sm:inline">{builder}</span>
+                              <span className="inline sm:hidden">{formatAddress(builder)}</span>
```
We could maybe inline the logic here : 

```ts
{builder ? builder?.slice(0, 5) + "..." + builder?.slice(-4) : "Address not found"}
```

Ahmedborwin commented:
Hey @technophile-04, 

Thanks for the detailed response. I will look at this in more detail this week and fix the issue + add the functionality I missed. 

I added the optimism check because I found that when I switched to a different chain, the read history hook did not work as expected. Not sure if the issue is with my code. I could revert the changes as requested and see if the issue persists and we can go from there?

technophile-04 commented:
> I found that when I switched to a different chain, the read history hook did not work as expected.

Ohh I see, its related to https://github.com/scaffold-eth/scaffold-eth-2/issues/633.... just pushed patch at [9f0dee0](https://github.com/BuidlGuidl/batch3.buidlguidl.com/pull/14/commits/9f0dee0831a38721569d255d52626d1e85cc7a85) since we are using only one chain and it should work fine now and can move ahead 🙌

technophile-04 commented:
Hey @Ahmedborwin, could you please resolve the merge conflicts, I think we are almost ready to merge 🙌

Ahmedborwin commented:
hey @technophile-04. I have been unable to get rid of the "hooks/scaffold-eth/useScaffoldEventSubscriber.ts(33,5): error TS2322: Type 'Abi' is not assignable to type 'InferContractAbi<Contract<TContractName>>'." even after following the instructions you sent and googling the issue. I tried to deal with the conflicts by initiating a PR from buidlguidl main to the issue 3 forked repo but that got messy and I did not understand what was going on anymore. Can you please clarify to me the best way to deal with the conflicts arising from the fact that the main repo has had PR requests merged since I created the issue 3 branch?

technophile-04 reviewed (approved):
> been unable to get rid of the "hooks/scaffold-eth/useScaffoldEventSubscriber.ts(33,5): error TS2322: Type 'Abi' is not assignable to type 'InferContractAbi<Contract>'."

I think husky(which we are using for pre-commit) might have some caching issue, could you please : 

1. Edit `useScaffoldEventSubscriber` maybe add a `abi: deployedContractData?.abi as any[],` at line 33 and save the file
2. Run `yarn next:check-types` ( this should give an error) 
3. Undo the changes put it back as it is, save it and again `yarn next:check-types` 

I think after this `yarn next:check-types` should (And should be able to commit) 

Nevertheless I have resolved conflicts at and merge `main`https://github.com/BuidlGuidl/batch3.buidlguidl.com/pull/14/commits/96b89f12ca8868dcd6a8ff78020eef33b6698fe8, but lets iterate over styles in future PR if we don't feel its good, Thanks @Ahmedborwin 🙌

Ahmedborwin commented:
Hey @technophile-04 - that fix worked thanks! 

Can I ask for future reference, what is the best way to deal with conflicts that arise from successful merges to the original repo while you work on a forked repo on a specific issue? Should I PR from original repo to the forked repo to deal with the conflicts then PR from my forked to main?
